### ISSUE TITLE
Suboptimal performance when using Walk::not

### ISSUE THREAD

**Issue by arlyon** [olson-sean-k/wax#49]
```rust
let glob = make_glob("crates/**/*.rs");
let walk = glob.walk("/path/to/dir").not(vec![make_glob("crates/**/target")]);
```

Consider the code below

https://github.com/olson-sean-k/wax/blob/1afcda8318201afc04ebed06fea907d54fc1bf8c/src/walk.rs#L749-L767

Given the globs above, the walker will end up traversing into every single `target` folder and never trigger a skip. The 'skip tree' optimisation will never trigger because it is only able to skip a folder after it is yielded to the negation via `self.input.next()`, and most `target` folders will a) not have rust files to yield and b) if they do, will not match the exclude case because they are a file and not the folder itself. That means that we will not correctly detect when we can `TreeIterator::skip_tree`. Seems like we need to invert the order here so that the `FilterTree` is wrapped by the `Walk` so that it may reject items first. Alternatively we could modify the `Walk` pattern to include the excluded globs, so that they are yielded to the `FilterTree` for exclusion but that seems a little inelegant...

I _believe_ that this will have no impact on the behaviour, but it will make walking with negations much more efficient.

**Comment by arlyon** [comment]
Note, it is possible to work around this by modifying the walk glob manually to match the behaviour I am describing, however globs are user supplied and so I cannot reasonably advise them to modify their globs.

```diff
- let glob = make_glob("crates/**/*.rs");
+ let glob = make_glob("crates/{target,**/*.rs}");
```

**Comment by olson-sean-k** [comment]
Patterns are never implicitly exhaustive. Just as `crates/**/target` will never positively match paths with names or components beyond `target` (for example, the path `crates/wax/target/debug/` does **not** match), a negation will not negatively match paths with names or components beyond `target` either. As a negation, `crates/**/target` only discards paths that terminate in a component matching `target` anywhere beneath a component matching `crates`. The negation `crates/**/target/**` is probably what you want here if I understand correctly. I would expect that pattern to avoid the spurious reads.

I think the most common kind of negations that folks want must terminate in a tree wildcard `**` like this. I've considered making this implicit, but that would force all negations to behave this way and I can imagine a need to avoid more specific paths, among other issues.

**Comment by arlyon** [comment]
Thanks for the quick reply! Using doublestar on the negation (`crates/**/target/**`) doesn't make a difference in my tests. I have attached some chrome traces of the various combinations. (not to scale, the importance is that big gap in the middle where we are neither yielding nor skipping folders).

`crates/**/*.rs - crates/**/target/**`

![image](https://github.com/olson-sean-k/wax/assets/8715352/174d49df-9043-492d-b616-52b4f4e76307)

`crates/**/*.rs - crates/**/target`

![image](https://github.com/olson-sean-k/wax/assets/8715352/f4e50c2b-5c7c-42b2-b06a-a19c9f9c0dec)

`crates/*/{target,**/*.rs} - crates/**/target/**`

![image](https://github.com/olson-sean-k/wax/assets/8715352/0acadff5-936c-49d1-8f36-5c255376f4f6)

`crates/*/{target,**/*.rs} - crates/**/target`

![image](https://github.com/olson-sean-k/wax/assets/8715352/4fd90b3e-59a1-46cf-996c-89bf2c261b61)

> The negation crates/**/target/** is probably what you want here if I understand correctly. I would expect that pattern to avoid the spurious reads.

You are correct, that does prevent `rs` files in `target` from being yielded (yay, part of the issue solved) but,  in terms of performance, the issue is that if the `Walk` doesn't yield the `target` folder to the `FilterTree` then we don't get the chance to skip it and so we end up traversing into the folder. If we end up yielding a `.rs` folder, the FilterTree will correctly ignore it but we still waste that time exploring a folder we don't need to.

**Comment by olson-sean-k** [comment]
> the issue is that if the `Walk` doesn't yield the `target` folder to the `FilterTree` then we don't get the chance to skip it

You're absolutely right! Filtering is not composed correctly and `FilterTree` only operates on the subset of items it receives from `Walk`. I'm working on a fix now.

**Comment by olson-sean-k** [comment]
After a bit of manual testing, I believe that I have a preliminary fix (currently on the [`filter`](https://github.com/olson-sean-k/wax/tree/filter) branch). It introduces `SeparatingFilter`s, which allow monotonic (one-way) filtering of a separation with special support for hierarchical tree iterators (like those reading directory trees from a file system) where all combinators can apply filter-mapping functions to the entire feed (both its filtrate and its residue).

`FileIterator` is now more general and calls to both `filter_entry` (previously `filter_tree`) and `not` can be chained against any `FileIterator`, including the newly introduced `WalkTree`, which does no matching nor filtering at all (unlike `WalkGlob`, which is returned by `Glob::walk`). This has introduced a significant amount of code, so I've reorganized the `walk` module and now expose it in the public API rather than re-exporting some of its items into the crate root.

This is a lot, but it preserves the current combinator API and its good ergonomics. 🎉

**Comment by arlyon** [comment]
Hello! Thank you for taking this on, you saved me a decent chunk of work :) I will pull down the branch and have a look.

**Comment by olson-sean-k** [comment]
I've landed a fix in 76afaa1 on `master`. :tada:

There are definitely some gaps in testing here, but [this unit test](https://github.com/olson-sean-k/wax/blob/76afaa120ea58f14f16fec28fd5d1de6737d3d7f/src/walk/mod.rs#L1130) _should_ detect the underlying problem that caused this issue.
